File: intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
'''
    a b c
          \
            h i j k .
          /
q d e f g
reverse bot

could put the references in a set for any one list, 
then iterate on the other until you find a node that is in the intersection

       c
          f
     d e 
b {d e f}
hA,hB (f null)
'''
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        b=set()
        while headB:
            b.add(headB)
            headB=headB.next
        while headA:
            if headA in b:
                return headA
            headA=headA.next
        return None

File: intersection-of-two-linked-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
//   9 9 9 3 2 6 4 1 8 4 5
//         i
//         j       
//   3 5 6 1 8 4 5

// 1. reverse the two lists!
// 2. then find *last* common node
// 3. then reverse both lists back!
// 4. return last
// complexity: same as before but space is constant now!

// sweep through headA and keep all in a hashmap
// then sweep through headB until you find something in the map... thats your answer!
// time: O(m + n)
// space: O(min(m,n))
// heada 1 2 3
// headb 2 3

// have = {1n 2n 3n}

type nothing struct {}
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    have := map[*ListNode]nothing {}
    for a := headA; a != nil; a = a.Next {
        have[a] = nothing{}
    }
    for b := headB; b != nil; b = b.Next {
        if _, ok := have[b]; ok { 
            return b 
        }
    }
    return nil
}